Library/Homebrew/cmd/mark.rb

```diff
@@ -0,0 +1,105 @@
+# typed: strict
+# frozen_string_literal: true
+
+require "abstract_command"
+require "formula"
+require "tab"
+
+module Homebrew
+  module Cmd
+    class Mark < AbstractCommand
+      cmd_args do
+        description <<~EOS
+          Mark or unmark <formula> as installed on request, or installed as
+          dependency. This can be useful if you want to control whether an
+          installed formula should be removed by `brew autoremove`.
+        EOS
+        switch "--installed-on-request",
+               description: "Mark <formula> as installed on request."
+        switch "--no-installed-on-request",
+               description: "Mark <formula> as not installed on request."
+        switch "--installed-as-dependency",
+               description: "Mark <formula> as installed as dependency."
+        switch "--no-installed-as-dependency",
+               description: "Mark <formula> as not installed as dependency."
```

Comment:
This is bit of a wider question but: what does this actually mean and why is it tracked separately? What does it mean for something to be neither on request and as a dependency, and what does it mean to be both on request and as a dependency?

Is it used anywhere? If it's unused then we probably don't need options to toggle it as there's no use case.

Reply:
> what does this actually mean and why is it tracked separately?

From what I can tell this represents whether the formula is installed as dependency *on its initial install*. After the initial install, the field is not updated even with explicit installs and reinstalls.

> What does it mean for something to be neither on request and as a dependency, and what does it mean to be both on request and as a dependency?

Neither on request nor as a dependency: I can't think of a situation where this combination makes sense, AFAICT.
Both on request and as a dependency: This is the state when formula `foo` is initially installed as a dependency of `bar`, but later you do e.g. `brew reinstall foo`.

> Is it used anywhere?

It is, just not in the autoremove logic. There are a few places where this field is accessed: https://github.com/search?q=repo%3AHomebrew%2Fbrew%20installed_as_dependency&type=code

But I don't think there's a legitimate use case for us to toggle this field, though. I'll remove the capability of manipulating this on my next push unless there's a reason not to.

Reply:
> There are a few places where this field is accessed: https://github.com/search?q=repo%3AHomebrew%2Fbrew%20installed_as_dependency&type=code

Apart from the printing logic, i.e. `brew leaves` and the recent `brew list` printing addition, it seems the use case is that if a formula was a non-leaf at any point in time in history (even if it has now since then become a leaf) then it will permanently enable falling back to building from source when there is no available bottle on an upgrade/reinstall.

It seems like a very odd separation that I don't think I understand yet.

Reply:
> From what I can tell this represents whether the formula is installed as dependency _on its initial install_. After the initial install, the field is not updated even with explicit installs and reinstalls.

On request is updated on reinstall, as dependency is not. I think this makes sense?

Reply:
> I think this makes sense?

I don't really have a strong opinion, but the following example does make me a bit confused:

Let's say `foo` is a dependency of `bar`.

Running `brew install foo` and `brew install bar` in order makes both not installed as dependency in the end.

Running `brew install bar` and `brew install foo` in order makes `foo` installed as dependency in the end. (`bar` is not, of course.)

I don't see why these two should be different, though?

Reply:
> I don't see why these two should be different, though?

Yeh, that's fair. Much as we rewrite `install on request` we should also perhaps rewrite `install as dependency` when installing something if it's in the dependency tree of something else? Not in this PR, though, but may help the metadata be better here.

Alternatively: we remove this field entirely as it doesn't seem terribly useful to us as-is.